### ISSUE TITLE
SPIR-V image type deduplication isn't right

### ISSUE THREAD

**Issue by jimblandy** [gfx-rs/naga#1305]
Naga generates bad SPIR-V for the following WGSL:

```
[[group(0), binding(0)]] var texture: texture_storage_2d<rgba32float, write>;
[[group(0), binding(0)]] var texture: texture_storage_2d<rgba32float, read>;
```

`spirv-val` rejects this with the error:
```
error: line 14: Duplicate non-aggregate type declarations are not allowed. Opcode: TypeImage id: 5
  %5 = OpTypeImage %float 2D 0 0 0 2 Rgba32f
```

Indeed, there are two identical `OpTypeImage` instructions in the generated SPIR-V:

```
          %3 = OpTypeImage %float 2D 0 0 0 2 Rgba32f
          %5 = OpTypeImage %float 2D 0 0 0 2 Rgba32f
```

This is because the SPIR-V backend deduplicates equal `LocalType` values, but `LocalType` equality requires `ImageClass::Storage::access`, to match, even though that has no influence on the `OpImageType` generated.

There may be other aspects of a Naga `Image` type that do not affect the generated `OpImageType`. Any two `Image` types that would generate the same `OpImageType` must have equal `LocalType` representations.

**Comment by jimblandy** [comment]
Except for this issue, I think we almost have a fix for #1257 with #1294 and #1306,

**Comment by kvark** [comment]
We should not have `StorageAccess` on `LocalType`. It's meant to be isomorphic to a subset of SPIR-V types.
